### PR TITLE
common/http: Remove chromium_url

### PR DESCRIPTION
**Pull request by dio**
Commit Message: This patch adds runtime feature flag to switch from chromium_url to url lib from googleurl. The new one is enabled by default.
Additional Description: This is provided as an alternative of #13808.
Risk Level: Medium
Testing: Existing, plus added unit tests for testing different runtime flags.
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
LGTM, if this can nuke `source/common/chromium_url` then this is a fairly clean approach (relatively, but the shim could still work..).

**Comment by dio** [comment]
Nuking `chromium_url` in this PR.

File: bazel/external/googleurl.patch
Comment (dio) [inline bazel/external/googleurl.patch:105]: Coverage and fuzz_coverage resolves more symbols from `url` lib, hence it is needed to include `url_canon_host.cc`.

Comment (dio) [reply bazel/external/googleurl.patch:105]: I feel the shimmed one is safer.

**Review by yanavlasov — APPROVED** [review]

**Review by yanavlasov — CHANGES_REQUESTED** [review]
Can you add a release note please?
/wait

File: test/common/http/path_utility_test.cc
Comment (yanavlasov) [inline test/common/http/path_utility_test.cc:31]: I think you need to use the `DEPRECATED_FEATURE_TEST` here as it will fail in one of the builds.

**Comment by dio** [comment]
Added an entry to release notes and made the "deprecated_path_canonicalizer.cc" to have 100% coverage.

**Review by yanavlasov — DISMISSED** [review]

File: source/common/http/path_utility.cc
Comment (htuch) [inline source/common/http/path_utility.cc:19]: Do we want to default true or false?

File: test/common/http/conn_manager_impl_test.cc
Comment (htuch) [inline test/common/http/conn_manager_impl_test.cc:443]: ?

File: bazel/external/googleurl.patch
Comment (htuch) [reply bazel/external/googleurl.patch:105]: @yanavlasov @dio thinking about this vs. the shim PR, I think ultimately, if we can upstream Chromium to give us an IDN-free optional build, that would be the ideal, since then we don't need to own a patch permanently. Shimming reduces the size of the patch, but in all cases with a patch we have a permanent patch ownership issue. LGTM modulo comments.

**Review by htuch — CHANGES_REQUESTED** [review]
/wait

File: source/common/http/path_utility.cc
Comment (dio) [reply source/common/http/path_utility.cc:19]: I think for now, until the URL parser is replaced, we set this "true". @yanavlasov WDYT?

File: test/common/http/path_utility_test.cc
Comment (dio) [reply test/common/http/path_utility_test.cc]: This is added. I actually not sure if we should go with "reloadable" flag or "deprecated_features" semantically. cc. @htuch.

File: test/common/http/conn_manager_impl_test.cc
Comment (dio) [reply test/common/http/conn_manager_impl_test.cc]: Sorry my bad, was trying to fix the typo. :D.

File: source/common/http/path_utility.cc
Comment (yanavlasov) [reply source/common/http/path_utility.cc:19]: If we make it `false` now we could get some test mileage, which could be useful for a change like this. This change it not really related to the URL parser, so I think `false` as default is ok.

File: test/common/http/path_utility_test.cc
Comment (yanavlasov) [reply test/common/http/path_utility_test.cc]: I think it should be `reloadable`. `deprecated_features` is for code changes that modify existing behavior.

File: bazel/external/googleurl.patch
Comment (yanavlasov) [reply bazel/external/googleurl.patch:109]: I recall this patch is needed to make fuzzing test happy. I was going to make another pass to see if tests can be fixed to get rid of the patch.

**Comment by yanavlasov** [comment]
LGTM, modulo changing the default.
/wait

File: bazel/external/googleurl.patch
Comment (dio) [reply bazel/external/googleurl.patch:109]: Yes, the last bit of this patch is related to fuzzing test. I agree we should take a deeper look on tests and modify it.

File: source/common/http/path_utility.cc
Comment (dio) [reply source/common/http/path_utility.cc]: Done. Changed using reloadable flag.

**Review by htuch — DISMISSED** [review]
LGTM; I'd like @alyssawilk to provide a final go/no-go before we merge, as another input on the runtime default direction, but good from my end. Thanks so much @dio for working through an area of large ambiguity and integration complexity, eliminating our fork here is a huge win.

**Comment by alyssawilk** [comment]
Generally we encourage new code paths on by default unless we think it's very high risk.  If this is roughly the same code, just a different source, I think on by default totally make sense :-)

**Review by yanavlasov — APPROVED** [review]

**Comment by dio** [comment]
@htuch @yanavlasov Sorry, fixing the conflicted current.rst dismissed the reviews.

**Review by yanavlasov — APPROVED** [review]

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by dio** [comment]
@htuch or @moderation we need lgtm deps? :)

**Comment by moderation** [comment]
/lgtm deps
Looks OK. Patch is large with only a categorization change for existing dep

**Pull request merged**
